config/rule.go

```diff
@@ -46,67 +49,179 @@ type Rule struct {
+	// Ensure the rule actually matches something.
+	//if r.Regex == nil && r.Path == nil {
+	//	return fmt.Errorf("%s: both |regex| and |path| are empty, this rule will have no effect", r.RuleID)
+	//}
+
+	// Ensure |secretGroup| works.
+	if r.Regex != nil && r.SecretGroup > r.Regex.NumSubexp() {
+		return fmt.Errorf("%s: invalid regex secret group %d, max regex secret group %d", r.RuleID, r.SecretGroup, r.Regex.NumSubexp())
+	}
+
+	return nil
 }
 
 type Verify struct {
 	HTTPVerb             string
 	URL                  string
 	Headers              map[string]string
-	ExpectedStatus       []string
+	ExpectedStatus       []int
 	ExpectedBodyContains []string
 
 	// RequiredIDs is a set of other rule IDs that must be present for verification.
+	initialized      bool
 	requiredIDs      map[string]struct{}
 	placeholderInUrl bool
 	staticHeaders    map[string]string
 	dynamicHeaders   map[string]string
 	//buildRequestFunc *func(req *http.Request, rule Rule, finding report.Finding, findingsByRuleID map[string][]report.Finding) func() string
 }
 
-func (v Verify) GetRequiredIDs() map[string]struct{} {
+func (v *Verify) GetRequiredIDs() map[string]struct{} {
 	return v.requiredIDs
 }
 
-func (v Verify) GetPlaceholderInUrl() bool {
+func (v *Verify) GetPlaceholderInUrl() bool {
 	return v.placeholderInUrl
 }
 
-func (v Verify) GetStaticHeaders() map[string]string {
+func (v *Verify) GetStaticHeaders() map[string]string {
 	return v.staticHeaders
 }
 
-func (v Verify) GetDynamicHeaders() map[string]string {
+func (v *Verify) GetDynamicHeaders() map[string]string {
 	return v.dynamicHeaders
 }
 
-// Validate guards against common misconfigurations.
-func (r Rule) Validate() error {
-	// Ensure |id| is present.
-	if strings.TrimSpace(r.RuleID) == "" {
-		// Try to provide helpful context, since |id| is empty.
-		var context string
-		if r.Regex != nil {
-			context = ", regex: " + r.Regex.String()
-		} else if r.Path != nil {
-			context = ", path: " + r.Path.String()
-		} else if r.Description != "" {
-			context = ", description: " + r.Description
+var (
+	verifyPlaceholderPat = regexp.MustCompile(`(?i)\${([a-z0-9\-]+)}`)
+	validMethods         = map[string]struct{}{
+		http.MethodGet:  {},
+		http.MethodHead: {},
+		http.MethodPost: {},
+	}
+)
+
+func (v *Verify) Validate(ruleID string) error {
+	if v.initialized {
+		return nil
+	}
+
+	// TODO: Check that there's some sort of substitution happening here.
+	v.requiredIDs = map[string]struct{}{}
+
+	// Validate verb.
+	if strings.TrimSpace(v.HTTPVerb) == "" {
+		return fmt.Errorf("%s: verify config does not contain a |httpVerb|", ruleID)
+	}
+	v.HTTPVerb = strings.ToUpper(v.HTTPVerb)
+	if _, ok := validMethods[v.HTTPVerb]; !ok {
+		return fmt.Errorf("%s: invalid HTTP verb: %s", ruleID, v.HTTPVerb)
+	}
+
+	// Parse URL.
+	if _, err := url.ParseRequestURI(v.URL); err != nil {
+		return fmt.Errorf("%s: invalid URL: %w", ruleID, err)
```

Comment:
**Question**: would this fail for urls with expanded strings? e.g, `example.com/${account-id}`?

Reply:
It won't fail unless the placeholder is in the hostname. I'm fine to remove it if that seems like a problem.

Reply:
seems like an unlikely case, probably fine for now